Snippet creation feature and steps.

landrysoules opened the pull request:
Unit tests to follow.
After snippet creation, user is now redirect to his snippets.
"My snippets" URL refactored in a more REST way => /users/:id/snippets.

landrysoules committed:
Snippet creation feature and steps.

Unit tests to follow.
After snippet creation, user is now redirect to his snippets.
"My snippets" URL refactored in a more REST way => /users/:id/snippets.

landrysoules committed:
Create snippet feature.

Rspec tests fixed (added authentication stub method call).

rastasheep commented on features/snippet.feature:
```diff
@@ -0,0 +1,7 @@
+
+    Scenario: Creating a snippet
+        Given I am authenticated
+        When I fill in the snippet creation form
```
Nice, maybe you could add one more step before this one where you'll go to the new snippet page (even if you have that in this step, just to be clear from where you're creating snippet).
P.S. generally we're using two spaces for indentation in all files

landrysoules replied:
You are absolutely right: it makes much more sense to specify the page where the action has to take place !
Concerning the indentation, I didn't notice that I had 3 spaces in my feature files... I will have to check how to change vim settings to play well with cucumber.

rastasheep replied:
For that purposes can recommend you this vim plugin https://github.com/tpope/vim-cucumber (it also includes some bonus stuff).

landrysoules replied:
Installed it with no success ;(

rastasheep commented on features/helpers.rb:
```diff
@@ -0,0 +1,23 @@
+
+module Helpers
+  def signup(user)
+    visit "/users/sign_out"
```
You probably do not need this first visit (in both helpers) or with that in place you're making sure that user is logged out ?

landrysoules replied:
Exactly: I wanted to make sure user is not signed when starting the test. I realize now that it seems  useless, however if I remove the visit to signout, the test fails...

rastasheep commented on features/step_definitions/snippet_steps.rb:
```diff
@@ -0,0 +1,23 @@
+  user = User.new email: "landry.soules@gmail.com", password: "1Password", username: "landry"
+  signup user
+  authenticate user
+  page.assert_text "Signed in successfully."
```
Use rspec expectations for checking stuff on the page like:

```
expect(page).to have_content("Signed in successfully.")
```

landrysoules replied:
Thanks, sometimes I have hard times figuring out where cucumber finishes and where rspec starts !

rastasheep replied:
It's not wrong but we're using rspec for expectations instead of capybara.

rastasheep commented on features/step_definitions/snippet_steps.rb:
```diff
@@ -0,0 +1,23 @@
+
+When(/^I fill in the snippet creation form$/) do
+  #TODO: check why cucumber doesn't recognize routing helpers
+  #visit new_snipet
```
Maybe you only need `_path` suffix to get this working like

```
visit new_snipet_path
```

landrysoules replied:
Thanks a lot ! I knew there must be some way to avoid writing down the url. 
By the way, it seems that visit_new_snipet_url also works.

rastasheep replied:
Yeah, both of those are URL helpers provided by Rails (http://guides.rubyonrails.org/routing.html#path-and-url-helpers), difference is in presence of host ("/snipets/new" vs "http://localhost:3000/"/snipets/new")

rastasheep commented on features/step_definitions/snippet_steps.rb:
```diff
@@ -0,0 +1,23 @@
+  visit "/snipets/new"
+  fill_in "snipet_title", :with => "my first snippet"
+  fill_in "snipet_content", :with => "ls"
+  click_button "submit_snipet"
```
You're not limited to ids only, you can also use classes or text which is on button (value) for submitting it.

rastasheep commented on features/step_definitions/snippet_steps.rb:
```diff
@@ -0,0 +1,23 @@
+end
+
+Then(/^I should see a success message$/) do
+  page.assert_text "Snipet was successfully created."
```
Same as above (expectations).

rastasheep commented on spec/controllers/snipets_controller_spec.rb:
```diff
@@ -57,6 +58,7 @@
 
   describe "GET #new" do
     it "assigns a new snipet as @snipet" do
+      sign_in  User.new(email: 'landry@spaceinvade.rs', password: '123456', username: 'landry')
```
You can pull this line to before filter, or better to shared context, and at the beginning of the spec file just include that context. Maybe you can use this as inspiration (https://www.relishapp.com/rspec/rspec-core/v/2-11/docs/example-groups/shared-context). Also user can be created with FactoryGirl, which allows you default values out of the box for  the tests. 

We'll check unit tests more after scenarios.

landrysoules replied:
Had very hard times trying to figure out how to use shared context, so in the end I just improved my code a little by using FactoryGirl for creating a snippet.
Shared contexts will definitely make a good candidate for a future workshop !

rastasheep commented on app/controllers/snipets_controller.rb:
```diff
@@ -46,11 +48,9 @@ def create
       if @snipet.save
-        format.html { redirect_to @snipet, notice: 'Snipet was successfully created.' }
-        format.json { render :show, status: :created, location: @snipet }
+        format.html { redirect_to my_snipets_url(current_user.id), notice: 'Snipet was successfully created.' }
```
When you have only html as format you do not need to specify it, you your code would look like:

```
...
@snipet.username = current_user.username
if @snipet.save
  redirect_to my_snipets_url(current_user.id), notice: 'Snipet was successfully created.' 
else
  render :new 
end
```

rastasheep commented:
:+1: Nice start :)

landrysoules commented:
Thank you so much Aleksandar, for your time and your priceless advices. It is exactly what I needed to be sure I don't follow the wrong path.
Cheers.

landrysoules committed:
Snippet creation - feature and steps.

Various code enhancements following Aleksandar's recommendations:
- use of RSpec to check content, in place of cucumber's methods where those were used for expectations
- use of routing helpers
- use of Factory Girl to generate a user before signin
- removal of useless json reponse handling in SnipetsController

rastasheep commented:
Let's add few other snippets scenarios, so we can merge this one, and we'll go back to specs later.

landrysoules committed:
Fixed rspec tests.
